File: src/utils/doseCalculator.js
// Utilidades de cálculo de dosis y de seguridad por edad.
// Uso educativo: todo resultado debe ser verificado por personal médico
// antes de administrar cualquier medicamento a un paciente pediátrico.

export function round(value, decimals = 1) {
  const factor = 10 ** decimals
  return Math.round(value * factor) / factor
}

// Determina el estado de seguridad por edad para un medicamento dado.
// Devuelve { level: 'unknown' | 'ok' | 'caution' | 'contraindicated', text }
export function getAgeSafety(med, ageMonths) {
  if (ageMonths === null || ageMonths === undefined || Number.isNaN(ageMonths)) {
    return {
      level: 'unknown',
      text: 'Edad no ingresada: verifique manualmente las restricciones por edad antes de usar este medicamento.',
    }
  }

  const flags = med.ageFlags || []
  // Se prioriza la restricción más severa que aplique a la edad indicada.
  const contraindicated = flags.find(
    (f) => f.level === 'contraindicated' && ageMonths < f.maxMonths
  )
  if (contraindicated) {
    return { level: 'contraindicated', text: contraindicated.text }
  }

  const caution = flags.find((f) => f.level === 'caution' && ageMonths < f.maxMonths)
  if (caution) {
    return { level: 'caution', text: caution.text }
  }

  return { level: 'ok', text: 'Sin restricción específica de edad para el rango indicado.' }
}

// Calcula la dosis para medicamentos de tipo "standard" (mg/kg con topes).
function computeStandardDose(med, weightKg) {
  const rawMin = weightKg * med.mgPerKgMin
  const rawMax = weightKg * med.mgPerKgMax

  const singleMin = med.maxSingleDoseMg ? Math.min(rawMin, med.maxSingleDoseMg) : rawMin
  const singleMax = med.maxSingleDoseMg ? Math.min(rawMax, med.maxSingleDoseMg) : rawMax

  const rawDaily = weightKg * med.maxDailyMgPerKg
  const dailyMax = med.maxDailyMg ? Math.min(rawDaily, med.maxDailyMg) : rawDaily

  const cappedBySingleMax = med.maxSingleDoseMg && rawMax > med.maxSingleDoseMg
  const cappedByDailyMax = med.maxDailyMg && rawDaily > med.maxDailyMg

  return {
    kind: 'standard',
    singleMin: round(singleMin),
    singleMax: round(singleMax),
    dailyMax: round(dailyMax),
    cappedBySingleMax,
    cappedByDailyMax,
    volumeMinMl: med.concentrationMgPerMl ? round(singleMin / med.concentrationMgPerMl) : null,
    volumeMaxMl: med.concentrationMgPerMl ? round(singleMax / med.concentrationMgPerMl) : null,
  }
}

// Calcula la dosis para el esquema especial de azitromicina (carga + mantenimiento).
function computeAzithromycinDose(med, weightKg) {
  const rawDayOne = weightKg * med.dayOneMgPerKg
  const rawMaintenance = weightKg * med.maintenanceMgPerKg

  const dayOne = Math.min(rawDayOne, med.dayOneMaxMg)
  const maintenance = Math.min(rawMaintenance, med.maintenanceMaxMg)

  return {
    kind: 'azithromycin',
    dayOne: round(dayOne),
    maintenance: round(maintenance),
    cappedDayOne: rawDayOne > med.dayOneMaxMg,
    cappedMaintenance: rawMaintenance > med.maintenanceMaxMg,
    volumeDayOneMl: med.concentrationMgPerMl ? round(dayOne / med.concentrationMgPerMl) : null,
    volumeMaintenanceMl: med.concentrationMgPerMl
      ? round(maintenance / med.concentrationMgPerMl)
      : null,
  }
}

// Calcula dosis por kg en unidades arbitrarias (mg, mcg, mL, mEq, U), sin
// conversión a mL de suspensión (no aplica a fármacos IV/IM/parenterales).
function computeWeightDose(med, weightKg) {
  const rawMin = weightKg * med.perKgMin
  const rawMax = weightKg * med.perKgMax

  const singleMin = med.maxSingle ? Math.min(rawMin, med.maxSingle) : rawMin
  const singleMax = med.maxSingle ? Math.min(rawMax, med.maxSingle) : rawMax
  const cappedBySingleMax = med.maxSingle && rawMax > med.maxSingle

  let dailyMax = null
  let cappedByDailyMax = false
  if (med.maxDailyPerKg) {
    const rawDaily = weightKg * med.maxDailyPerKg
    dailyMax = med.maxDailyMax ? Math.min(rawDaily, med.maxDailyMax) : rawDaily
    cappedByDailyMax = Boolean(med.maxDailyMax && rawDaily > med.maxDailyMax)
    dailyMax = round(dailyMax, 2)
  }

  return {
    kind: 'weightDose',
    unit: med.unit,
    singleMin: round(singleMin, 2),
    singleMax: round(singleMax, 2),
    cappedBySingleMax,
    dailyMax,
    cappedByDailyMax,
  }
}

// Calcula una tasa de infusión (por kg/min o kg/h) para el peso del paciente.
// No se calcula mL/h: depende de la concentración de preparación de cada
// institución, y asumir una lo convertiría en un cálculo no verificable.
function computeInfusionDose(med, weightKg) {
  const rawMin = weightKg * med.perKgMin
  const rawMax = weightKg * med.perKgMax
  const cappedByMaxRate = med.maxRate && rawMax > med.maxRate
  const rateMin = med.maxRate ? Math.min(rawMin, med.maxRate) : rawMin
  const rateMax = med.maxRate ? Math.min(rawMax, med.maxRate) : rawMax

  return {
    kind: 'infusion',
    unit: med.unit,
    timeUnit: med.timeUnit,
    rateMin: round(rateMin, 2),
    rateMax: round(rateMax, 2),
    cappedByMaxRate,
  }
}

// Selecciona el texto de dosis fija correspondiente al tramo de edad del
// paciente (en meses). No implica multiplicación por peso ni edad.
function computeAgeTierDose(med, ageMonths) {
  if (ageMonths === null || ageMonths === undefined || Number.isNaN(ageMonths)) {
    return { kind: 'ageTier', doseText: null, needsAge: true }
  }
  const tier = med.tiers.find((t) => ageMonths < t.maxMonths) || med.tiers[med.tiers.length - 1]
  return { kind: 'ageTier', doseText: tier.doseText, needsAge: false }
}

// Selecciona el texto de dosis fija correspondiente al tramo de peso del
// paciente (en kg).
function computeWeightTierDose(med, weightKg) {
  const tier = med.tiers.find((t) => weightKg < t.maxKg) || med.tiers[med.tiers.length - 1]
  return { kind: 'weightTier', doseText: tier.doseText }
}

// Dosis plana, sin cálculo por peso ni edad (la fuente no la expresa por kg).
function computeFixedDose(med) {
  return { kind: 'fixed', doseText: med.doseText }
}

// Selecciona la dosis mg/kg (u otra unidad) y la frecuencia según el tramo de
// edad gestacional al nacer (semanas) y edad postnatal (días) del neonato —
// ambas dimensiones a la vez, ya que la dosificación neonatal depende de las
// dos. Se prueban los tramos en el orden declarado y se usa el primero cuyo
// límite superior de AMBAS dimensiones sea mayor que los valores del
// paciente (mismo criterio "primer match" que ageTier/weightTier, pero en 2D).
function computeNeonatalTierDose(med, weightKg, gestationalWeeks, postnatalDays) {
  if (
    gestationalWeeks === null ||
    gestationalWeeks === undefined ||
    Number.isNaN(gestationalWeeks) ||
    postnatalDays === null ||
    postnatalDays === undefined ||
    Number.isNaN(postnatalDays)
  ) {
    return { kind: 'neonatalTier', needsInput: true }
  }

  const tiers = med.neonatalTiers
  const tier =
    tiers.find((t) => gestationalWeeks < t.maxGestationalWeeks && postnatalDays < t.maxPostnatalDays) ||
    tiers[tiers.length - 1]

  const hasWeight = typeof weightKg === 'number' && weightKg > 0
  const rawMin = hasWeight ? weightKg * tier.perKgMin : null
  const rawMax = hasWeight ? weightKg * tier.perKgMax : null
  const doseMin = rawMin !== null && tier.maxSingle ? Math.min(rawMin, tier.maxSingle) : rawMin
  const doseMax = rawMax !== null && tier.maxSingle ? Math.min(rawMax, tier.maxSingle) : rawMax
  const capped = Boolean(hasWeight && tier.maxSingle && rawMax > tier.maxSingle)

  return {
    kind: 'neonatalTier',
    needsInput: false,
    needsWeight: !hasWeight,
    doseMin: doseMin !== null ? round(doseMin, 2) : null,
    doseMax: doseMax !== null ? round(doseMax, 2) : null,
    unit: med.unit,
    frequencyText: tier.frequencyText,
    capped,
  }
}

// Igual que computeNeonatalTierDose, pero para fármacos cuya tabla original
// tiera por PESO ACTUAL (no edad gestacional) + edad postnatal — se usa el
// mismo peso ya ingresado para el cálculo de la dosis y para elegir el tramo.
function computeNeonatalWeightTierDose(med, weightKg, postnatalDays) {
  const hasWeight = typeof weightKg === 'number' && weightKg > 0
  if (!hasWeight || postnatalDays === null || postnatalDays === undefined || Number.isNaN(postnatalDays)) {
    return { kind: 'neonatalWeightTier', needsInput: true, needsWeight: !hasWeight }
  }

  const tiers = med.neonatalWeightTiers
  const tier =
    tiers.find((t) => weightKg < t.maxWeightKg && postnatalDays < t.maxPostnatalDays) ||
    tiers[tiers.length - 1]

  const rawMin = weightKg * tier.perKgMin
  const rawMax = weightKg * tier.perKgMax
  const doseMin = tier.maxSingle ? Math.min(rawMin, tier.maxSingle) : rawMin
  const doseMax = tier.maxSingle ? Math.min(rawMax, tier.maxSingle) : rawMax
  const capped = Boolean(tier.maxSingle && rawMax > tier.maxSingle)

  return {
    kind: 'neonatalWeightTier',
    needsInput: false,
    needsWeight: false,
    doseMin: round(doseMin, 2),
    doseMax: round(doseMax, 2),
    unit: med.unit,
    frequencyText: tier.frequencyText,
    capped,
  }
}

export function computeDose(med, weightKg, ageMonths, gestationalWeeks, postnatalDays) {
  if (med.doseType === 'azithromycin') {
    return computeAzithromycinDose(med, weightKg)
  }
  if (med.doseType === 'weightDose') {
    return computeWeightDose(med, weightKg)
  }
  if (med.doseType === 'infusion') {
    return computeInfusionDose(med, weightKg)
  }
  if (med.doseType === 'ageTier') {
    return computeAgeTierDose(med, ageMonths)
  }
  if (med.doseType === 'weightTier') {
    return computeWeightTierDose(med, weightKg)
  }
  if (med.doseType === 'fixed') {
    return computeFixedDose(med)
  }
  if (med.doseType === 'neonatalTier') {
    return computeNeonatalTierDose(med, weightKg, gestationalWeeks, postnatalDays)
  }
  if (med.doseType === 'neonatalWeightTier') {
    return computeNeonatalWeightTierDose(med, weightKg, postnatalDays)
  }
  return computeStandardDose(med, weightKg)
}

File: windows-app/PedMed.Core/Services/DoseCalculator.cs
using PedMed.Core.Models;

namespace PedMed.Core.Services;

/// <summary>
/// Dose calculation and age-safety logic. A faithful C# port of
/// src/utils/doseCalculator.js from the React web app — keep both in sync.
/// Educational use only: every result must be verified by a clinician before
/// administering any medication to a pediatric patient.
/// </summary>
public static class DoseCalculator
{
    public static double Round(double value, int decimals = 1)
    {
        var factor = Math.Pow(10, decimals);
        return Math.Round(value * factor, MidpointRounding.AwayFromZero) / factor;
    }

    /// <summary>Formatea el rango de dosis por kg usado en el cálculo, ej. "10 – 15 mg/kg".</summary>
    private static string FormatPerKgFormula(double min, double max, string unit)
    {
        return Math.Abs(min - max) < 0.0001
            ? $"{FmtNum(min)} {unit}"
            : $"{FmtNum(min)} – {FmtNum(max)} {unit}";
    }

    private static string FmtNum(double v) =>
        v.ToString("0.##", System.Globalization.CultureInfo.InvariantCulture);

    /// <summary>Determines the age-safety badge/level for a medication given the patient's age in months.</summary>
    public static AgeSafetyResult GetAgeSafety(Medication med, double? ageMonths)
    {
        if (ageMonths is null || double.IsNaN(ageMonths.Value))
        {
            return new AgeSafetyResult(
                SafetyLevel.Unknown,
                "Edad no ingresada: verifique manualmente las restricciones por edad antes de usar este medicamento.");
        }

        var age = ageMonths.Value;

        // Se prioriza la restricción más severa que aplique a la edad indicada.
        var contraindicated = med.AgeFlags.FirstOrDefault(
            f => f.Level == SafetyLevel.Contraindicated && age < f.MaxMonths);
        if (contraindicated is not null)
        {
            return new AgeSafetyResult(SafetyLevel.Contraindicated, contraindicated.Text);
        }

        var caution = med.AgeFlags.FirstOrDefault(
            f => f.Level == SafetyLevel.Caution && age < f.MaxMonths);
        if (caution is not null)
        {
            return new AgeSafetyResult(SafetyLevel.Caution, caution.Text);
        }

        return new AgeSafetyResult(SafetyLevel.Ok, "Sin restricción específica de edad para el rango indicado.");
    }

    private static DoseResult ComputeStandardDose(Medication med, double weightKg)
    {
        var rawMin = weightKg * med.MgPerKgMin!.Value;
        var rawMax = weightKg * med.MgPerKgMax!.Value;

        var singleMin = med.MaxSingleDoseMg is { } maxSingle ? Math.Min(rawMin, maxSingle) : rawMin;
        var singleMax = med.MaxSingleDoseMg is { } maxSingle2 ? Math.Min(rawMax, maxSingle2) : rawMax;

        var rawDaily = weightKg * med.MaxDailyMgPerKg!.Value;
        var dailyMax = med.MaxDailyMg is { } maxDaily ? Math.Min(rawDaily, maxDaily) : rawDaily;

        var cappedBySingleMax = med.MaxSingleDoseMg is { } ms && rawMax > ms;
        var cappedByDailyMax = med.MaxDailyMg is { } md && rawDaily > md;

        return new DoseResult
        {
            Kind = DoseType.Standard,
            Formula = FormatPerKgFormula(med.MgPerKgMin!.Value, med.MgPerKgMax!.Value, "mg/kg"),
            SingleMin = Round(singleMin),
            SingleMax = Round(singleMax),
            DailyMax = Round(dailyMax),
            CappedBySingleMax = cappedBySingleMax,
            CappedByDailyMax = cappedByDailyMax,
            VolumeMinMl = med.ConcentrationMgPerMl is { } c1 ? Round(singleMin / c1) : null,
            VolumeMaxMl = med.ConcentrationMgPerMl is { } c2 ? Round(singleMax / c2) : null,
        };
    }

    private static DoseResult ComputeAzithromycinDose(Medication med, double weightKg)
    {
        var rawDayOne = weightKg * med.DayOneMgPerKg!.Value;
        var rawMaintenance = weightKg * med.MaintenanceMgPerKg!.Value;

        var dayOne = Math.Min(rawDayOne, med.DayOneMaxMg!.Value);
        var maintenance = Math.Min(rawMaintenance, med.MaintenanceMaxMg!.Value);

        return new DoseResult
        {
            Kind = DoseType.Azithromycin,
            DayOneFormula = FormatPerKgFormula(med.DayOneMgPerKg!.Value, med.DayOneMgPerKg!.Value, "mg/kg"),
            MaintenanceFormula = FormatPerKgFormula(med.MaintenanceMgPerKg!.Value, med.MaintenanceMgPerKg!.Value, "mg/kg"),
            DayOne = Round(dayOne),
            Maintenance = Round(maintenance),
            CappedDayOne = rawDayOne > med.DayOneMaxMg.Value,
            CappedMaintenance = rawMaintenance > med.MaintenanceMaxMg.Value,
            VolumeDayOneMl = med.ConcentrationMgPerMl is { } c1 ? Round(dayOne / c1) : null,
            VolumeMaintenanceMl = med.ConcentrationMgPerMl is { } c2 ? Round(maintenance / c2) : null,
        };
    }

    private static DoseResult ComputeWeightDose(Medication med, double weightKg)
    {
        var rawMin = weightKg * med.PerKgMin!.Value;
        var rawMax = weightKg * med.PerKgMax!.Value;

        var singleMin = med.MaxSingle is { } maxSingle ? Math.Min(rawMin, maxSingle) : rawMin;
        var singleMax = med.MaxSingle is { } maxSingle2 ? Math.Min(rawMax, maxSingle2) : rawMax;
        var cappedBySingleMax = med.MaxSingle is { } ms && rawMax > ms;

        double? dailyMax = null;
        var cappedByDailyMax = false;
        if (med.MaxDailyPerKg is { } maxDailyPerKg)
        {
            var rawDaily = weightKg * maxDailyPerKg;
            var daily = med.MaxDailyMax is { } maxDailyMax ? Math.Min(rawDaily, maxDailyMax) : rawDaily;
            cappedByDailyMax = med.MaxDailyMax is { } mdm && rawDaily > mdm;
            dailyMax = Round(daily, 2);
        }

        return new DoseResult
        {
            Kind = DoseType.WeightDose,
            Formula = FormatPerKgFormula(med.PerKgMin!.Value, med.PerKgMax!.Value, $"{med.Unit}/kg"),
            Unit = med.Unit,
            SingleMin = Round(singleMin, 2),
            SingleMax = Round(singleMax, 2),
            CappedBySingleMax = cappedBySingleMax,
            DailyMax = dailyMax,
            CappedByDailyMax = cappedByDailyMax,
        };
    }

    private static DoseResult ComputeInfusionDose(Medication med, double weightKg)
    {
        var rawMin = weightKg * med.PerKgMin!.Value;
        var rawMax = weightKg * med.PerKgMax!.Value;
        var cappedByMaxRate = med.MaxRate is { } mr && rawMax > mr;
        var rateMin = med.MaxRate is { } maxRate ? Math.Min(rawMin, maxRate) : rawMin;
        var rateMax = med.MaxRate is { } maxRate2 ? Math.Min(rawMax, maxRate2) : rawMax;

        return new DoseResult
        {
            Kind = DoseType.Infusion,
            Formula = FormatPerKgFormula(med.PerKgMin!.Value, med.PerKgMax!.Value, $"{med.Unit}/kg/{med.TimeUnit}"),
            Unit = med.Unit,
            TimeUnit = med.TimeUnit,
            RateMin = Round(rateMin, 2),
            RateMax = Round(rateMax, 2),
            CappedByMaxRate = cappedByMaxRate,
        };
    }

    private static DoseResult ComputeAgeTierDose(Medication med, double? ageMonths)
    {
        if (ageMonths is null || double.IsNaN(ageMonths.Value))
        {
            return new DoseResult { Kind = DoseType.AgeTier, DoseText = null, NeedsAge = true };
        }

        var age = ageMonths.Value;
        var tiers = med.AgeTiers!;
        var tier = tiers.FirstOrDefault(t => age < t.MaxMonths) ?? tiers[^1];
        return new DoseResult { Kind = DoseType.AgeTier, DoseText = tier.DoseText, NeedsAge = false };
    }

    private static DoseResult ComputeWeightTierDose(Medication med, double weightKg)
    {
        var tiers = med.WeightTiers!;
        var tier = tiers.FirstOrDefault(t => weightKg < t.MaxKg) ?? tiers[^1];
        return new DoseResult { Kind = DoseType.WeightTier, DoseText = tier.DoseText };
    }

    private static DoseResult ComputeFixedDose(Medication med)
    {
        return new DoseResult { Kind = DoseType.Fixed, DoseText = med.DoseText };
    }

    /// <summary>
    /// Selects the mg/kg (or other unit) dose and frequency according to the tier whose
    /// gestational age at birth (weeks) and postnatal age (days) both exceed the patient's
    /// values — both dimensions at once, since neonatal dosing depends on both. Tiers are
    /// tried in declaration order and the first one whose upper bound in BOTH dimensions
    /// exceeds the patient's values is used (same "first match" rule as AgeTier/WeightTier,
    /// extended to 2D).
    /// </summary>
    private static DoseResult ComputeNeonatalTierDose(Medication med, double? weightKg, double? gestationalWeeks, double? postnatalDays)
    {
        if (gestationalWeeks is null || double.IsNaN(gestationalWeeks.Value) ||
            postnatalDays is null || double.IsNaN(postnatalDays.Value))
        {
            return new DoseResult { Kind = DoseType.NeonatalTier, NeedsInput = true };
        }

        var tiers = med.NeonatalTiers!;
        var tier = tiers.FirstOrDefault(t => gestationalWeeks.Value < t.MaxGestationalWeeks && postnatalDays.Value < t.MaxPostnatalDays)
                   ?? tiers[^1];

        var hasWeight = weightKg is > 0;
        double? rawMin = hasWeight ? weightKg!.Value * tier.PerKgMin : null;
        double? rawMax = hasWeight ? weightKg!.Value * tier.PerKgMax : null;
        var doseMin = rawMin is { } rMin && tier.MaxSingle is { } msMin ? Math.Min(rMin, msMin) : rawMin;
        var doseMax = rawMax is { } rMax && tier.MaxSingle is { } msMax ? Math.Min(rMax, msMax) : rawMax;
        var capped = hasWeight && tier.MaxSingle is { } ms2 && rawMax > ms2;

        return new DoseResult
        {
            Kind = DoseType.NeonatalTier,
            NeedsInput = false,
            NeedsWeight = !hasWeight,
            Formula = FormatPerKgFormula(tier.PerKgMin, tier.PerKgMax, $"{med.Unit}/kg"),
            DoseMin = doseMin is { } dMin ? Round(dMin, 2) : null,
            DoseMax = doseMax is { } dMax ? Round(dMax, 2) : null,
            Unit = med.Unit,
            FrequencyText = tier.FrequencyText,
            Capped = capped,
        };
    }

    /// <summary>
    /// Same as <see cref="ComputeNeonatalTierDose"/>, but for medications whose source
    /// table tiers by CURRENT weight (not gestational age) — reuses the already-entered
    /// weight both to select the tier and to compute the dose.
    /// </summary>
    private static DoseResult ComputeNeonatalWeightTierDose(Medication med, double? weightKg, double? postnatalDays)
    {
        var hasWeight = weightKg is > 0;
        if (!hasWeight || postnatalDays is null || double.IsNaN(postnatalDays.Value))
        {
            return new DoseResult { Kind = DoseType.NeonatalWeightTier, NeedsInput = true, NeedsWeight = !hasWeight };
        }

        var tiers = med.NeonatalWeightTiers!;
        var tier = tiers.FirstOrDefault(t => weightKg!.Value < t.MaxWeightKg && postnatalDays.Value < t.MaxPostnatalDays)
                   ?? tiers[^1];

        var rawMin = weightKg!.Value * tier.PerKgMin;
        var rawMax = weightKg.Value * tier.PerKgMax;
        var doseMin = tier.MaxSingle is { } msMin ? Math.Min(rawMin, msMin) : rawMin;
        var doseMax = tier.MaxSingle is { } msMax ? Math.Min(rawMax, msMax) : rawMax;
        var capped = tier.MaxSingle is { } ms2 && rawMax > ms2;

        return new DoseResult
        {
            Kind = DoseType.NeonatalWeightTier,
            NeedsInput = false,
            NeedsWeight = false,
            Formula = FormatPerKgFormula(tier.PerKgMin, tier.PerKgMax, $"{med.Unit}/kg"),
            DoseMin = Round(doseMin, 2),
            DoseMax = Round(doseMax, 2),
            Unit = med.Unit,
            FrequencyText = tier.FrequencyText,
            Capped = capped,
        };
    }

    /// <summary>
    /// Computes the dose result for a medication given the patient's weight (kg) and
    /// optional age (months). Age is only required for DoseType.AgeTier. Gestational
    /// weeks and postnatal days are only required for DoseType.NeonatalTier /
    /// DoseType.NeonatalWeightTier — pass weightKg as 0 (or any non-positive value) when
    /// the weight has not been entered yet, mirroring the "no weight yet" sentinel already
    /// used for AgeTier/Fixed by the view-model.
    /// </summary>
    public static DoseResult Compute(
        Medication med,
        double weightKg,
        double? ageMonths = null,
        double? gestationalWeeks = null,
        double? postnatalDays = null)
    {
        return med.DoseType switch
        {
            DoseType.Azithromycin => ComputeAzithromycinDose(med, weightKg),
            DoseType.WeightDose => ComputeWeightDose(med, weightKg),
            DoseType.Infusion => ComputeInfusionDose(med, weightKg),
            DoseType.AgeTier => ComputeAgeTierDose(med, ageMonths),
            DoseType.WeightTier => ComputeWeightTierDose(med, weightKg),
            DoseType.Fixed => ComputeFixedDose(med),
            DoseType.NeonatalTier => ComputeNeonatalTierDose(med, weightKg > 0 ? weightKg : null, gestationalWeeks, postnatalDays),
            DoseType.NeonatalWeightTier => ComputeNeonatalWeightTierDose(med, weightKg > 0 ? weightKg : null, postnatalDays),
            _ => ComputeStandardDose(med, weightKg),
        };
    }
}
